graphix/opengraph.py

```diff
@@ -0,0 +1,162 @@
+"""Provides a class for open graphs."""
+
+from __future__ import annotations
+
+import math
+from typing import TYPE_CHECKING, NamedTuple
+
+import networkx as nx
+
+from graphix.generator import generate_from_graph
+
+if TYPE_CHECKING:
+    from collections.abc import Mapping
+
+    from graphix.pattern import Pattern
+    from graphix.pauli import Plane
+
+
+class Measurement(NamedTuple):
+    """An MBQC measurement.
+
+    :param angle: the angle of the measurement. Should be between [0, 2)
+    :param plane: the measurement plane
+    """
+
+    angle: float
+    plane: Plane
+
+    def isclose(self, other: Measurement, rel_tol=1e-09, abs_tol=0.0) -> bool:
+        """Compares if two measurements have the same plane and their angles
+        are close.
+
+        Example
+        -------
+        >>> from graphix.opengraph import Measurement
+        >>> from graphix.pauli import Plane
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        True
+        >>> Measurement(0.0, Plane.XY).isclose(Measurement(0.0, Plane.YZ))
+        False
+        >>> Measurement(0.1, Plane.XY).isclose(Measurement(0.0, Plane.XY))
+        False
+        """
+        return math.isclose(self.angle, other.angle, rel_tol=rel_tol, abs_tol=abs_tol) and self.plane == other.plane
+
+
+class OpenGraph:
```

Comment:
Because our measurements are stored as floats, there may be OpenGraphs where `__eq__` returns true because `Measurement.isclose` returns true for all the measurements, but the hashes won't be equal because the floats are not exactly the same.

I'm not sure if this may lead to issues in future so I'm not sure if it is the best to implement a naive hash function now unless we think hashing will be required

Reply:
I see, understood why `__hash__` is missing in the current implementation.
By the way, I strongly believe that `__eq__` should test strict equality, rather than logical equality defined by `math.isclose`.
I remember there was a discussion about that.
Where was it?

Reply:
We could instead turn the `OpenGraph.__eq__` method into another `isclose` method?
How does that sound?

Testing strict equality of open graphs wouldn't work well due to the angles being implemented as floats

Reply:
> I remember there was a discussion about that... Where was it?

We already had a discussion about `__eq__`/`isclose` in the PR about Pauli measures: https://github.com/TeamGraphix/graphix/pull/189#issuecomment-2267505193

Reply:
> Testing strict equality of open graphs wouldn't work well due to the angles being implemented as floats

I understand it, but still we had better not hack `__eq__`.
It is really confusing for users and also unconventional.

Reply:
I'm not familiar with the changes made in that PR. To keep consistent with the design decisions in that PR, is it better to make `__eq__` into an `isclose` method for open graphs? What is the way forward here?

Reply:
Okay I've converted this method to be `isclose`.
OpenGraph should get a default implementation of `__eq__` from the dataclass decorator which will compare the attributes as if it was a tuple.

Reply:
@EarlMilktea If all looks good to you, can you resolve this conversation and see if there is anything else to change before we merge?